### DISCUSSION TITLE
Auto assignation of custom objects attributes

### DISCUSSION THREAD

**Discussion by Aluriak** [textX/textX#351] [Ideas]
Hi !

I'm working with textx these days, it's so cool, i love it.

Now, i hate doing the same thing each time, so, instead of writing a custom class like this:

    class Model:
        def __init__(self, commands):
            self.commands = commands
        ...

I prefer to write this:

    class Model:
        ...

But that sounds impossible, without (aggressive or implicit) changes in textx itself.
I therefore write this instead:

    class Model(metaclass=model_class):
        ...

With the following definition somewhere before:

    def model_class(name: str, bases: tuple, attrs: dict) -> type:
        """Metaclass to automatically build the __init__ to get the properties,
        and register the class for metamodel
        """
        if "__init__" not in attrs:

            def __init__(self, **kwargs):
                for field, value in kwargs.items():
                    setattr(self, field, value)

            attrs["__init__"] = __init__
        cls = type(name, bases, attrs)
        model_class.classes.append(cls)
        return cls
    model_class.classes = []

(i leave the `model_class.classes` thing here if it inspire someone, but i doubt it to be useful in this context ; feel free to ignore that bit)

Questions:

- is my metaclass correct ? (it is in all the cases i used it)
- whatabout including it in textx, something like `textx.regular_model_metaclass`, with a proper documentation ?

**Comment by goto40** [comment]
Sounds useful. I had the same 'poblem'. However, I am not a big metaclass expert.

It like the idea.

**Comment by igordejanovic** [comment]
@Aluriak Yup. Very nice idea. Could be generally useful *but* we have to support Python 2.x also. Thus, no type hinting and metaclass spec. differs between versions (I didn't do a deep dive at this point so the difference might be only at the point of usage). Thus, we should take that into account and cover it with tests that pass for all supported version. And of course, cover the new feature in the docs :) 

If you have time to take on that it would be great. Let's leave this open as a new feature request anyway.

There is a nice library called `six` that ease the cross-version support. I had some [slight metaclass  juggling in another project](https://github.com/igordejanovic/parglare/blob/master/parglare/grammar.py#L1621) using `six` and it works very well so we could either introduce `six` or maybe just do `if` guards on python versions (I always prefer to keep dependencies at minimum).

**Comment by Aluriak** [comment]
@goto40 my definitive read-every-time-i-use-it guide is [this stack answer](https://stackoverflow.com/a/6581949/3077939).

@igordejanovic i use [that cheatsheet](https://python-future.org/compatible_idioms.html#metaclasses), when working on compat. I understand the problem ; my (future) PR will adress that.

**Comment by Aluriak** [comment]
Ok, major revelation (coming between two dreams this night), the metaclass thing is unnecessary if you only need the automatic attribute initialization:

    class ModelClass(object):  # py2 compat
        def __init__(self, **kwargs):
            for name, value in kwargs.items()
                 setattr(self, name, value)

    class Model(ModelClass):
        ...

Also, now i'm thinking on it, automatic collection of classes can be done with a namespace:

    class ClassesForGrammar:

        class A(ModelClass):
            ...
        class B(ModelClass):
            ...
        class C(ModelClass):
            ...


    classes = tuple(cls for name, cls in vars(ClassesForGrammar).items()
                    if not name.startswith('_'))
    print(classes)

I would go for metaclass because that's fully integrated and transparent, but in the same time, thinking about code that would have multiple grammars to handle, the second solution makes sens.

I will go for the second, then.

**Comment by Aluriak** [comment]
@igordejanovic Just understood that your metaclass snippet was from parglare. I read your blog post few times ago. Really hyped by the incoming parser :)

**Comment by igordejanovic** [comment]
@Aluriak Yes, it sounds better and more straightforward than using meta-classes.

**Comment by wryun** [comment]
@Aluriak the incredibly dirty thing I'm doing is letting textx create its default classes then automagically monkey patching methods on afterwards based on naming conventions. One probably shouldn't do this.
